padl/transforms.py

```diff
@@ -2184,24 +2425,78 @@ class _ItemGetter:
     :param default: The default value to fall back to in case of exception.
     """
 
-    def __init__(self, samples, transform, exception=None, default=None):
+    def __init__(self, samples, transform, entire_transform, exception=None, default=None):
         self.samples = samples
         self.transform = transform
+        self.entire_transform = entire_transform
+        self.exception = exception
+        self.default = default
+
+    def __getitem__(self, item):
+        if self.exception:
+            try:
+                return item, self.transform(self.samples[item])
+            except self.exception:
+                return item, self.default
+        try:
+            return item, self.transform(self.samples[item])
+        except Exception as err:
+            is_entire_transform = str(self.entire_transform) == \
+                                  str(self.entire_transform.pd_preprocess)
+            self.entire_transform._pd_trace_error(
+                self.entire_transform.pd_preprocess._pd_get_target_stage_idx(is_entire_transform),
+                [self.samples[item]]
+            )
+            raise err
+
+    def __len__(self):
+        return len(self.samples)
+
+
+class _BatchGetter:
```

Comment:
@eramdiaz As you mentioned before we could probably use `_ItemGetter` instead of `_BatchGetter`. Instead of adding an option to `_ItemGetter` you could also pass the `Identity()` as the `transform` input argument for `_ItemGetter`.

Reply:
yes, but there is also one more thing that could overload the class, in `_BatchGetter` we are returning the index on a list, like 
`return [item], self.samples[item]`, while in `_ItemGetter` we are returning just the index `return item, self.samples[item]`

Reply:
In that case adding a simple check if `item` is a `list` or not would work, right?
```python
if isinstance(item, int):
     item = [item]
```

Reply:
In both cases `item` is an integer (because in the `use_preprocess` case we call later a `DataLoader` that makes the list of indices and the `args`). What could work using the current code would be call `_ItemGetter` with `Identity` and then `DataLoader` with `batch_size=1'.

Reply:
This solution is causing issues. It is changing datatypes because of internal `DataLoader` working mode and making a test fail.

Reply:
@wuhu @sjrl

Reply:
Sounds good, having a separate Getter is a good solution then.